File: client-rust/src/transaction.rs
use mercury_lib::transaction::{SignFirstRequestPayload, PartialSignatureMsg1, PartialSignatureRequestPayload, PartialSignatureResponsePayload};
use anyhow::Result;
use secp256k1_zkp::musig::MusigPartialSignature;
use crate::client_config::ClientConfig;

/// This function gets the server public nonce from the statechain entity.
pub async fn sign_first(client_config: &ClientConfig, sign_first_request_payload: &SignFirstRequestPayload) -> Result<String> {

    let endpoint = client_config.statechain_entity.clone();
    let path = "sign/first";

    let client: reqwest::Client = reqwest::Client::new();
    let request = client.post(&format!("{}/{}", endpoint, path));

    let value = request.json(&sign_first_request_payload).send().await?.text().await?;

    let sign_first_response_payload: mercury_lib::transaction::SignFirstResponsePayload = serde_json::from_str(value.as_str())?;

    let mut server_pubnonce_hex = sign_first_response_payload.server_pubnonce.to_string();

    if server_pubnonce_hex.starts_with("0x") {
        server_pubnonce_hex = server_pubnonce_hex[2..].to_string();
    }

    Ok(server_pubnonce_hex)
}

pub async fn get_server_partial_sig(client_config: &ClientConfig, partial_sig_request: &PartialSignatureRequestPayload) -> Result<MusigPartialSignature> {
    let endpoint = client_config.statechain_entity.clone();
    let path = "sign/second";

    let client: reqwest::Client = reqwest::Client::new();
    let request = client.post(&format!("{}/{}", endpoint, path));

    let value = request.json(&partial_sig_request).send().await?.text().await?;

    let response: PartialSignatureResponsePayload = serde_json::from_str(value.as_str())?;

    let mut server_partial_sig_hex = response.partial_sig.to_string();

    if server_partial_sig_hex.starts_with("0x") {
        server_partial_sig_hex = server_partial_sig_hex[2..].to_string();
    }

    let server_partial_sig_bytes = hex::decode(server_partial_sig_hex)?;

    let server_partial_sig = MusigPartialSignature::from_slice(server_partial_sig_bytes.as_slice())?;

    Ok(server_partial_sig)
}

File: clients/mercury-layer-base/src/wallet/transaction.js

import axios from 'axios';
import config from 'config';
import SocksProxyAgentLib from 'socks-proxy-agent';

const SocksProxyAgent = SocksProxyAgentLib.SocksProxyAgent;

const signFirst = async (signFirstRequestPayload) => {

    const statechain_entity_url = config.get('statechainEntity');
    const path = "sign/first";
    const url = statechain_entity_url + '/' + path;

    const torProxy = config.get('torProxy');

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }
    
    let response = await axios.post(url, signFirstRequestPayload, socksAgent);

    let server_pubnonce_hex = response.data.server_pubnonce;

    return server_pubnonce_hex;
}

const signSecond = async (partialSigRequest) => {

    const statechain_entity_url = config.get('statechainEntity');
    const path = "sign/second";
    const url = statechain_entity_url + '/' + path;

    const torProxy = config.get('torProxy');

    let socksAgent = undefined;

    if (torProxy) {
        socksAgent = { httpAgent: new SocksProxyAgent(torProxy) };
    }

    let response = await axios.post(url, partialSigRequest, socksAgent);

    let server_partial_sig_hex = response.data.partial_sig;

    return server_partial_sig_hex;
}

export default { signFirst, signSecond };